private/pkg/manifest/manifest_test.go

```diff
@@ -0,0 +1,194 @@
+// Copyright 2020-2022 Buf Technologies, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package manifest
+
+import (
+	"bytes"
+	"context"
+	"errors"
+	"fmt"
+	"strings"
+	"testing"
+	"testing/iotest"
+
+	"github.com/bufbuild/buf/private/pkg/storage/storagemem"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+	"golang.org/x/crypto/sha3"
+)
+
+func Example() {
+	ctx := context.Background()
+	bucket, _ := storagemem.NewReadBucket(
+		map[string][]byte{
+			"foo": []byte("bar"),
+		},
+	)
+	manifest, _ := NewManifestFromBucket(ctx, bucket)
+	digest, _ := manifest.GetDigest("foo")
+	fmt.Printf("digest[:16]: %s\n", digest.Hex()[:16])
+	path, _ := manifest.GetPath(digest)
+	fmt.Printf("path at digest: %s\n", path)
+	// Output:
+	// digest[:16]: a15163728ed24e1c
+	// path at digest: foo
+}
+
+func mkdigest(content []byte) *Digest {
+	hash := sha3.NewShake256()
+	if _, err := hash.Write(content); err != nil {
+		panic(err)
+	}
+	digest := make([]byte, 64)
+	if _, err := hash.Read(digest); err != nil {
+		panic(err)
+	}
+	return NewDigestFromBytes("shake256", digest)
+}
+
+func TestRoundTripManifest(t *testing.T) {
+	// read a manifest using the unmarshaling method
+	null := mkdigest([]byte{})
+	var manifestBuilder bytes.Buffer
+	for i := 0; i < 2; i++ {
+		fmt.Fprintf(&manifestBuilder, "%s  null%d\n", null, i)
```

Comment:
iiuc this test, this will result in this content:

```
  null0
  null1
```

If so I have 2 Qs:
- It has 2 spaces, does the `strings.Cut` result in `hash:""`, `path:" null0"`?
- The hash doesn't contain `:`, why is not triggering an error?

Reply:
Nevermind, I'm being silly, `mkdigest([]byte{})` will generate a valid hash.

Reply:
It'll result in a manifest with `shake256:xxxx    null<n>` because that's what `Digest.String` returns, which is what `null` is. It is a tad hard to read in a test though. I'll make it a little more explicit and less short-handy.

> It has 2 spaces, does the strings.Cut result in hash:"", path:" null0"?

Two spaces and strings.Cut takes that as the separator. It'll result in path as "null0" (without spaces). Same goes with why ":" is contained within but you cannot easily see it here.